dnf-repoconfig-daemon/repoconf.cpp

```diff
@@ -0,0 +1,169 @@
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Library General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
+ */
+
+#include "configuration.hpp"
+#include "repoconf.hpp"
+
+#include <iostream>
+#include <sdbus-c++/sdbus-c++.h>
+#include <string>
+
+
+RepoConf::RepoConf(sdbus::IConnection &connection) : connection(connection)
+{
+    dbus_register_methods();
+}
+
+void RepoConf::dbus_register_methods()
+{
+    const std::string objectPath = "/org/rpm/dnf/v0/rpm/RepoConf";
+    const std::string interfaceName = "org.rpm.dnf.v0.rpm.RepoConf";
+
+    dbus_object = sdbus::createObject(connection, objectPath);
+    dbus_object->registerMethod(interfaceName, "list", "as", "aa{sv}", [this](sdbus::MethodCall call) -> void {this->list(call);});
+    dbus_object->registerMethod(interfaceName, "get", "s", "a{sv}", [this](sdbus::MethodCall call) -> void {this->get(call);});
+    dbus_object->registerMethod(interfaceName, "enable", "as", "as", [this](sdbus::MethodCall call) -> void {this->enable_disable(call, true);});
+    dbus_object->registerMethod(interfaceName, "disable", "as", "as", [this](sdbus::MethodCall call) -> void {this->enable_disable(call, false);});
+    dbus_object->finishRegistration();
+}
+
+bool RepoConf::check_authorization(const std::string &actionid, const std::string &sender)
+{
+    // create proxy for PolicyKit1 object
+    const std::string destinationName = "org.freedesktop.PolicyKit1";
+    const std::string objectPath = "/org/freedesktop/PolicyKit1/Authority";
+    const std::string interfaceName = "org.freedesktop.PolicyKit1.Authority";
+    auto polkit_proxy = sdbus::createProxy(connection, destinationName, objectPath);
+    polkit_proxy->finishRegistration();
+ 
+    // call CheckAuthorization method
+    sdbus::Struct<bool, bool, std::map<std::string, std::string>> auth_result;
+    sdbus::Struct<std::string, KeyValueMap> subject {"system-bus-name", {{"name", sender}}};
+    std::map<std::string, std::string> details {};
+    uint flags = 0;
+    std::string cancelation_id = "";
+    polkit_proxy->callMethod("CheckAuthorization").onInterface(interfaceName).withArguments(subject, actionid, details, flags, cancelation_id).storeResultsTo(auth_result);
+
+    // get results
+    bool res_is_authorized = std::get<0>(auth_result);
+    //bool res_is_challenge = std::get<1>(auth_result);
+    //std::map<std::string, std::string> res_details = std::get<2>(auth_result);
+
+    return res_is_authorized;
+}
+
+KeyValueMapList RepoConf::repo_list(const std::vector<std::string> &ids)
```

Comment:
Is it ever a use case that yum.repos.d/ files contain secrets and thus are not world-readable? If so, then listing repos should be restricted as well (polkit).